app/src/org/commcare/tasks/ProcessAndSendTask.java

```diff
@@ -174,60 +177,68 @@ private boolean checkFormRecordStatus(FormRecord[] records)
 
             //If the form is complete, but unprocessed, process it.
             if (FormRecord.STATUS_COMPLETE.equals(record.getStatus())) {
+                SQLiteDatabase db =
+                        CommCareApplication.instance().getUserDbHandle();
+                db.beginTransaction();
                 try {
                     records[i] = processor.process(record);
+                    db.setTransactionSuccessful();
                 } catch (InvalidStructureException | XmlPullParserException |
                         UnfullfilledRequirementsException e) {
-                    handleExceptionFromFormProcessing(record, e);
+                    db.endTransaction();
+                    records[i] = handleExceptionFromFormProcessing(record, e);
                     needToSendLogs = true;
                 } catch (FileNotFoundException e) {
+                    db.endTransaction();
                     if (CommCareApplication.instance().isStorageAvailable()) {
                         //If storage is available generally, this is a bug in the app design
                         Logger.log(LogTypes.TYPE_ERROR_DESIGN,
                                 "Removing form record because file was missing|" + getExceptionText(e));
                         record.logPendingDeletion(TAG,
                                 "the xml submission file associated with the record could not be found");
                         FormRecordCleanupTask.wipeRecord(c, record);
+                        records[i] = FormRecord.StandInForDeletedRecord();
                     } else {
                         CommCareApplication.notificationManager().reportNotificationMessage(
                                 NotificationMessageFactory.message(ProcessIssues.StorageRemoved), true);
                         //Otherwise, the SD card just got removed, and we need to bail anyway.
                         throw e;
                     }
                 } catch (IOException e) {
+                    db.endTransaction();
                     Logger.log(LogTypes.TYPE_ERROR_WORKFLOW, "IO Issues processing a form. " +
                             "Tentatively not removing in case they are resolvable|" + getExceptionText(e));
+                } catch (Exception e) {

```

Comment:
If you want this to execute for ~all code paths this should be a `Throwable`, since this will miss `Error` throws. I put an alternative possibility above that I think might be a bit cleaner (and less risky that someone forgets to end a transaction in a new catch())

Reply:
Good point, changed